### DISCUSSION TITLE
Deterministic behaviour for TASing

### DISCUSSION THREAD

**Discussion by clementgallet** [86Box/86Box#2913] [Ideas]
**Is your feature request related to a problem? Please describe.**
As 86Box is supported on Linux using SDL2, it is possible to make [TASes](https://en.wikipedia.org/wiki/Tool-assisted_speedrun) through the use of [libTAS](https://github.com/clementgallet/libTAS/). However, 86Box is currently not running deterministically regarding inputs handling, rendering and emulation. Indeed, the three tasks are done in separate threads and running concurrently. To make inputs replay deterministic, we would need them to run sequentially, always in the same order.

**Describe the solution you'd like**
To support this, we would need to remove all threading, and run sequentially `inputs handling > emulation step (10 ms) > rendering`, and also be sure to render on every loop, even if screen buffer did not change.
I'd like to know if you would consider this change to be merged into the main project? If so, how would you prefer that I implement it: compiler option? emulator option? This would only concerns the Linux SDL2 backend.

**Describe alternatives you've considered**
We could also just fork and modify on our end, but it is cleaner if we can use upstream version to TAS, instead of relying on a fork that we would have to keep updated

**Comment by Cacodemon345** [comment]
An emulator option is unsuitable for our purposes because it would become a major performance hit, which makes a compiler option more suitable.

I would also usually advise against the compile-time option as well, but the SDL2 backend is considered minimal enough that a compile-time option wouldn't hurt unless it also involves large rewrites of other emulator code, in which case I would instead advise to fork and modify on your end.

By the way, maybe we could accept your patches provided that it is done with mutexes (as you mentioned in the libTAS issue), doesn't involve large rewrites and is provided as a compile-time option, since, like I said, the SDL2 backend is minimal enough.

**Comment by OBattler** [comment]
The problem there is not even threading - the renderer thread is still being informed by a call from the CPU thread and that also happens if nothing has changed. The sequentiality is broken by the fact that it's not a strict input -> emulate -> render - the CPU sees the input whenever the appropriate timer has been polled and the resulting interrupt processed.

The one way I can see for TAS'ing, is saving the state at the beginning of every CPU frame (= clock speed in Hz / 100 cycles, so 10 ms) and adding an option to rewind to a previous frame.

**Comment by clementgallet** [comment]
Thanks, this could be indeed achieved with extra mutexes for much smaller code rewrite. However, in addition to SDL backend code, rewrites would be needed in `src/video/video.c` and in video drivers that use threading, such as voodoo for example.

Saving the emulator state is already handled successfully by libTAS, it uses screen rendering calls to delimit frames. TASing is achieved already on PCem with a [fork for determinism](https://github.com/TASEmulators/pcem) (with the modifications talked here, but with less threading instead of more sync functions), and people already [made TASes](https://tasvideos.org/7425S) with this setup.

**Comment by Cacodemon345** [comment]
I don't think you need to rewrite the blitter, just use condition variables for synchronization with main thread.
